File: 75_Problems.js
// ISSUES
//Trees: Maximum Path Sum, Serialize Binary Tree

//Two Sum

//Hash (45% Speed, 54% space)
var twoSum = function (nums, target) {
    let targets = {};
    for (let i = 0; i < nums.length; i++) {
        let subTarget = target - nums[i];
        if (targets[subTarget] !== undefined) {
            return [targets[subTarget], i];
        } else {
            targets[nums[i]] = i
        }
    }
    return []
};

//Two for loops (12% speed, 97% space)
var twoSum = function (nums, target) {
    for (let i = 0; i < nums.length - 1; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) return [i, j];
        }
    }

};

// Buy and Sell Stock

// One iteration through (13% speed, 74% space)
var maxProfit = function (prices) {
    let maxReturn = 0;
    let currentMin = prices[0];
    for (let i = 1; i < prices.length; i++) {
        let currentDay = prices[i];
        if (currentDay - currentMin > maxReturn) {
            maxReturn = (currentDay - currentMin)
        }
        currentMin = Math.min(currentDay, currentMin)
    }
    return maxReturn;
};

//Contains Duplicates
// One iteration through (22% speed, 73% space)
var containsDuplicate = function (nums) {
    let values = new Set();
    for (let i = 0; i < nums.length; i++) {
        if (values.has(nums[i])) {
            return true
        } else {
            values.add(nums[i])
        }
    }
    return false
};

//Sort first (92% speed, 80.5% space)
var containsDuplicate = function (nums) {
    let sorted = nums.sort((a, b) => {
        return a - b
    })
    for (let i = 1; i < nums.length; i++) {
        if (nums[i - 1] === nums[i]) return true;
    }
    return false
};

//Product of array except self

//two nested loops, failed on time
var productExceptSelf = function (nums) {
    let newArr = new Array(nums.length).fill(1);
    for (let i = 0; i < nums.length; i++) {
        let val = nums[i];
        for (let j = 0; j < nums.length; j++) {
            if (j !== i) {
                newArr[i] *= nums[j]
            }
        }
    }
    return newArr
};

//two independent loops (time 18%, space 40%)
var productExceptSelf = function (nums) {
    let lefts = new Array(nums.length)
    lefts[0] = 1;
    for (let i = 1; i < nums.length; i++) {
        lefts[i] = lefts[i - 1] * nums[i - 1]
    }

    let right = 1;
    for (let j = nums.length - 1; j >= 0; j--) {
        lefts[j] *= right;
        right *= nums[j]
    }
    return lefts
};


//Max Sum Subarray
var maxSubArray = function (nums) {
    let currMax = nums[0];
    let testMax = nums[0];
    for (let i = 1; i < nums.length; i++) {
        if (testMax <= 0 && nums[i] > testMax) {
            testMax = nums[i]
            currMax = currMax > testMax ? currMax : testMax;
        } else {
            testMax += nums[i];
            currMax = testMax > currMax ? testMax : currMax;
        }
        console.log('currMax', currMax)
    }
    return currMax;
};

// Time 60%, space 29%
var maxSubArray = function (nums) {
    let currentMax = nums[0];
    let currentSum = nums[0];
    for (let i = 1; i < nums.length; i++) {
        if (currentSum < 0) {
            currentSum = nums[i]
        } else {
            currentSum += nums[i]
        }

        currentMax = currentSum > currentMax ? currentSum : currentMax;
    }
    return currentMax
};

//Max Product Subarray
// Time 16%, Space 73%
var maxProduct = function (nums) {
    let currMax = Number.NEGATIVE_INFINITY;
    let currProduct = 1;
    let currProduct2 = 1;
    let foundFalse = false;
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] == 0) {
            currMax = currMax > 0 ? currMax : 0;
            currProduct = 1;
            foundFalse = false;
            currProduct2 = 1;

        } else {
            currProduct *= nums[i]
            if (foundFalse) {
                currProduct2 *= nums[i]
                currMax = currProduct2 > currMax ? currProduct2 : currMax
            }
            else if (nums[i] < 0) {
                foundFalse = true;
            }

            currMax = currProduct > currMax ? currProduct : currMax
        }
    }
    return currMax;
};

//Time 58%, space 39%
var maxProduct = function (nums) {
    currMin = nums[0];
    currMax = nums[0];
    answer = nums[0];
    let min, max;
    for (let i = 1; i < nums.length; i++) {
        let val = nums[i];
        min = currMin;
        max = currMax;
        currMin = Math.min(val, min * val, max * val);
        currMax = Math.max(val, min * val, max * val);
        answer = Math.max(currMax, answer)
    }
    return answer
};

//Search for Min in rotated sorted array
var findMin = function (nums) {
    console.log(nums)
    if (nums.length <= 2) return Math.min(...nums);
    let middle = Math.floor(nums.length / 2);
    let end = nums[nums.length - 1];
    if (nums[middle] < nums[middle - 1]) {
        return nums[middle]
    } else if (nums[middle] < end) {
        return findMin(nums.slice(0, middle))
    } else {
        return findMin(nums.slice(middle + 1))
    }
};

//recursive
var findMin = function (nums) {
    if (nums.length === 1 || nums[0] < nums[nums.length - 1]) return nums[0]
    let middle = Math.floor(nums.length / 2);
    let start = nums[0];
    if (nums[middle] < nums[middle - 1]) {
        return nums[middle]
    } else if (nums[middle] < start) {
        return findMin(nums.slice(0, middle))
    } else {
        return findMin(nums.slice(middle + 1))
    }
};

//Two pointers
var findMin = function (nums) {
    let left = 0;
    let right = nums.length - 1;
    while (left < right) {
        if (nums[left] < nums[right]) return nums[left];
        let middle = Math.floor((left + right) / 2);
        if (nums[middle] > nums[right]) {
            left = middle + 1
        } else {
            right = middle
        }
    }
    return nums[left]
};


//DYNAMIC PROGRAMMING

// Number of Steps = Memoization
var climbStairs = function (n) {
    let memo = {}
    function climb(number) {
        if (number == 1) return 1;
        if (number == 0) return 1;
        let firstClimb = memo[number - 1] || climb(number - 1);
        let secondClimb = memo[number - 2] || climb(number - 2)
        memo[number - 1] = firstClimb;
        memo[number - 2] = secondClimb;
        return firstClimb + secondClimb
    }
    return climb(n)
};

//Coin Change

//Memo with recursion
var coinChange = function (coinsTotal, amountTotal) {
    let memo = {};
    function makeChange(coins, amount) {
        if (amount < 0) return -1;
        if (amount === 0) return 0;
        if (coins.includes(amount)) return 1;
        let ways = [];
        let result;
        for (let i = 0; i < coins.length; i++) {
            if (memo[amount - coins[i]] != undefined) {
                result = memo[amount - coins[i]]
            } else {
                result = makeChange(coins, amount - coins[i]);
                memo[amount - coins[i]] = result
            }
            if (result !== -1) {
                ways.push(result + 1)
            }
        }
        if (ways.length == 0) {
            return -1;
        } else {
            return Math.min(...ways)
        }
    }
    result = makeChange(coinsTotal, amountTotal)
    console.log(memo)
    return result;
};

//Longest Increasing Subsequence
var lengthOfLIS = function (nums) {
    let tails = [nums[0]];
    let value, placed;
    for (let i = 1; i < nums.length; i++) {
        value = nums[i];
        placed = false;
        for (let j = 0; j < tails.length; j++) {
            if (value <= tails[j]) {
                tails[j] = value;
                placed = true;
                break;
            }
        }
        if (!placed) tails.push(value)
    }
    console.log(tails)
    return tails.length
};

//Word Break
var wordBreak = function (s, wordDict) {
    let wordSet = new Set(wordDict)
    let results = new Array(s.length + 1).fill(false)
    results[0] = true;
    for (let i = 0; i < s.length; i++) {
        if (!results[i]) continue;
        for (let j = 1; j < s.length + 1; j++) {
            if (wordSet.has(s.slice(i, j))) {
                results[j] = true;
            }
        }
    }
    return results[results.length - 1]
};

// Intervals

var insert = function (intervals, newInterval) {
    if (intervals.length === 0) {
        return [newInterval]
    }
    let start;
    let inserted = false;
    for (let i = 0; i < intervals.length; i++) {
        if (intervals[i][1] >= newInterval[0] && newInterval[1] >= intervals[i][0]) {
            intervals[i] = [Math.min(intervals[i][0], newInterval[0]), Math.max(intervals[i][1], newInterval[1])]
            start = i;
            inserted = true;
            break;
        }
    }
    if (!inserted) {
        intervals.push(newInterval)
        intervals.sort((a, b) => {
            return a[0] - b[0]
        })
        return intervals
    }
    console.log(intervals)
    let newSet = [intervals[0]];
    for (let i = 0; i < intervals.length; i++) {
        let lastInterval = newSet[newSet.length - 1];
        if (intervals[i][0] <= lastInterval[1]) {
            newSet[newSet.length - 1] = [lastInterval[0], Math.max(intervals[i][1], lastInterval[1])]
        } else {
            newSet.push(intervals[i])
        }
    }
    return newSet;
};

//TREES
var maxDepth = function (root) {
    let currentMaxDepth = 0;
    function getDepth(node, value) {
        if (node == null) {
            currentMaxDepth = value > currentMaxDepth ? value : currentMaxDepth;
        } else {
            getDepth(node.left, value + 1)
            getDepth(node.right, value + 1)
        }
    }

    getDepth(root, 0);
    return currentMaxDepth;
};

var isSameTree = function (p, q) {
    let pQueue = [p]
    let qQueue = [q]
    let pNode, qNode, pVal, qVal
    while (pQueue.length > 0) {
        pNode = pQueue.shift();
        qNode = qQueue.shift();
        if (qNode == null && pNode == null) {
            continue
        }
        pVal = pNode ? pNode.val : null;
        qVal = qNode ? qNode.val : null;
        if (pVal == qVal) {
            if (pVal != null) pQueue.push(pNode.left, pNode.right)
            if (qVal != null) qQueue.push(qNode.left, qNode.right)
        } else {
            return false
        }

    }
    return pQueue.length == qQueue.length
};

var invertTree = function (root) {
    if (root == null) return null;
    let right = invertTree(root.right)
    let left = invertTree(root.left)
    root.left = right
    root.right = left
    return root
};

var levelOrder = function (root) {
    if (!root) return [];
    let result = [];
    let queue = [root];
    let currNode;
    let nextQueue = [];
    let currResult = [];
    while (queue.length > 0) {
        queue.forEach((node) => {
            currResult.push(node.val)
            if (node.left) {
                nextQueue.push(node.left)
            }
            if (node.right) nextQueue.push(node.right)
        })
        result.push(currResult)
        queue = nextQueue;
        nextQueue = [];
        currResult = [];
    }
    return result;
};

var isSubtree = function (root, subRoot) {
    function isSame(node1, node2) {
        if (node1 == null && node2 == null) return true;
        let oneVal = node1 ? node1.val : null;
        let twoVal = node2 ? node2.val : null;
        if (oneVal == twoVal) {
            return isSame(node1.left, node2.left) && isSame(node1.right, node2.right)
        } else {
            return false;
        }
    }

    let queue = [root];
    let currNode;
    while (queue.length > 0) {
        currNode = queue.shift();
        if (currNode.val == subRoot.val) {
            if (isSame(currNode, subRoot)) {
                return true
            }
        }
        if (currNode.left) queue.push(currNode.left);
        if (currNode.right) queue.push(currNode.right)
    }
    return false;
};

function preorderTraversal(root) {
    let values = [];
    function getValue(node) {
        if (node == null) return;

        values.push(node.val)
        getValue(node.left)
        getValue(node.right)
    }
    getValue(root)
    return values;
}

function inorderTraversal(root) {
    let values = [];
    function getValue(node) {
        if (node == null) return;
        
        getValue(node.left)
        values.push(node.val)
        getValue(node.right)
    }
    getValue(root)
    return values;
}

function postorderTraversal(root) {
    let values = [];
    function getValue(node) {
        if (node == null) return;

        getValue(node.left)
        getValue(node.right)
        values.push(node.val)
    }
    getValue(root)
    return values;
}

//build tree from pre and inorder arrays
var buildTree = function (preorder, inorder) {
    function makeTree(pre, inor) {
        if (pre.length == 0) {
            return null
        }
        let head = new TreeNode(pre[0]);
        let rootPos = inor.indexOf(pre[0]);
        head.left = makeTree(pre.slice(1, 1 + rootPos), inor.slice(0, rootPos))
        head.right = makeTree(pre.slice(rootPos + 1), inor.slice(rootPos + 1))
        return head;
    }
    return makeTree(preorder, inorder)

};

//Valid BST
var isValidBST = function (root) {
    function checkNode(node, minimum, maximum) {
        if (node == null) return true;
        if (node.val <= minimum || node.val >= maximum) {
            return false;
        }
        return (checkNode(node.left, minimum, node.val) && checkNode(node.right, node.val, maximum))
    }
    return checkNode(root);
};

//Kth smallest
var kthSmallest = function (root, k) {
    let values = [];
    function traverse(node) {
        if (node == null) return;
        traverse(node.left)
        if (values.length >= k) return;
        values.push(node.val)
        if (values.length >= k) return;
        traverse(node.right)
    }
    traverse(root)
    return values[k - 1]
};

//Lowest common Ancestor
var lowestCommonAncestor = function (root, p, q) {
    if (!root) return null;
    let queue = [[root, []]]
    let pAncestors, qAncestors, currNode, prev;
    let ancestors;
    while (!pAncestors || !qAncestors) {

        [currNode, prev] = queue.shift()
        ancestors = prev.concat(currNode)
        if (currNode == p) {
            pAncestors = ancestors
        }
        if (currNode == q) {
            qAncestors = ancestors
        }
        if (currNode.left) queue.push([currNode.left, ancestors])
        if (currNode.right) queue.push([currNode.right, ancestors])
    }

    let minLength = Math.min(pAncestors.length, qAncestors.length)
    for (let i = 0; i < minLength; i++) {
        if (pAncestors[i] !== qAncestors[i]) {
            return pAncestors[i - 1] ? pAncestors[i - 1] : pAncestors[i]
        }
        if (i == minLength - 1) {
            return pAncestors[i]
        }
    }
};

var lowestCommonAncestor = function (root, p, q) {
    while (root) {
        if (p.val < root.val && q.val < root.val) {
            root = root.left
        } else if (p.val > root.val && q.val > root.val) {
            root = root.right
        } else {
            break
        }
    }
    return root;
}

File: 75_Problems.rb
def two_sum(nums, target)
    numsHash = {};
    nums.each_with_index do |num, idx|
        if (numsHash.include?(target - num))
            return [numsHash[target - num], idx]
        else
            numsHash[num] = idx
        end         
    end
end

def max_profit(prices)
    curr_max_profit = 0
    curr_cheapest_price = prices[0]
    for i in 1...prices.length
        curr_max_profit = curr_max_profit > (prices[i] - curr_cheapest_price) ? curr_max_profit :  prices[i] - curr_cheapest_price
        curr_cheapest_price = prices[i] < curr_cheapest_price ? prices[i] : curr_cheapest_price
    end
    curr_max_profit
end

require 'set'
def contains_duplicate(nums)
    numsSet = Set.new()
    nums.each do |num|
        return true if numsSet.include?(num) 
        numsSet.add(num)
    end
    false
end

def product_except_self(nums)
    left_result = [1]
    for i in (1...nums.length)
        left_result << left_result[i - 1] * nums[i - 1] 
    end
    right_result = Array.new(nums.length)
    j = nums.length - 1
    right_result[j] = 1
    j-=1
    while (j >= 0)
        right_result[j] = nums[j + 1] * right_result[j + 1]
        j-=1
    end
    right_result.zip(left_result).map { |a,b| a*b}
end

def max_sub_array(nums)
    curr_max_sum = nums[0]
    curr_sum = nums[0]
    for i in (1...nums.length)
        curr_sum = curr_sum + nums[i] >= nums[i] ? curr_sum + nums[i] : nums[i]
        curr_max_sum = [curr_sum, curr_max_sum].max
    end
    curr_max_sum
end

def max_product(nums)
    curr_max = nums[0]
    curr_min = nums[0]
    total_max = nums[0]
    for i in 1...nums.length
        curr_value = nums[i]
        test_max = [curr_max * curr_value, curr_min*curr_value, curr_value].max
        test_min = [curr_max * curr_value, curr_min*curr_value, curr_value].min
        curr_max = test_max
        curr_min = test_min
        total_max = [curr_max, total_max].max
    end
    total_max
end

#Find Min sorted Array (logn)
def find_min(nums)
    return nums[0] if nums.length == 1
    start = 0
    last = nums.length - 1
    return nums[start] if (nums[start] < nums[last])
    while (start <= last)
        middle = (start + last) / 2
        return nums[middle+1] if (nums[middle] > nums[middle + 1])
        return nums[middle] if (nums[middle] < nums[middle - 1]) 
        curr_test = nums[middle]
        if (curr_test > nums[0])
            start = middle + 1
        elsif (curr_test < nums[start])
            last = middle -1
        end
    end
    nums[start]
end

# Search in sorted array
def search(nums, target)
    left = 0
    right = nums.length - 1
    while (left <= right)
       middle = (left + right) / 2
        return middle if (nums[middle]) == target
        if (nums[left] <= nums[middle])
            if (target >= nums[left] && target < nums[middle])
                right = middle - 1
            else
                left = middle + 1
            end
        else
            if (target <= nums[right] && target > nums[middle])
                left = middle + 1
            else
                right = middle - 1
            end
        end
    end
        nums[left] == target ? left: -1;
end

require 'set'
def three_sum(nums)
    s = Set.new()
    return [] if nums.length < 3
    result = [];
    for i in (0...nums.length-2)
        for j  in (1...nums.length-1)
           for m in (2...nums.length)
              if (nums[i] + nums[j] + nums[m] == 0 && i != j && j != m && m != i) 
                  string = [nums[i], nums[j],nums[m]].sort {|a,b| a-b}.join('')
                  if (s.include?(string))
                      next
                  else
                        result << [nums[i], nums[j], nums[m]] 
                      s.add(string)
                  end
              end
           end
        end
    end
    result.sort
end

def three_sum(nums)
    sorted = nums.sort
    result = []
    print sorted
    for i in (0...nums.length - 2)
       next if (sorted[i] > 0)
       next if (i > 0 && sorted[i] == sorted[i- 1])
        left = i + 1
        right = sorted.length - 1
        while (left < right)
            sum = sorted[i] + sorted[left] + sorted[right]
            if (sum > 0)
                right-=1
            elsif (sum < 0)
                left+=1
            else
                result << [sorted[i], sorted[left], sorted[right]]
            
                while (left < right && sorted[left] == sorted[left + 1])
                    left+=1
                end
                while (right > left && sorted[right] == sorted[right - 1])
                    right-=1
                end
                left += 1
                right -= 1
            end
        end
    end
    result
end

# Container with most water
def max_area(height)
    left = 0
    right = height.length - 1
    curr_max = 0
    while (left < right)
       curr_min = [height[left], height[right]].min
        curr_total = curr_min*(right - left)
        curr_max = curr_total > curr_max ? curr_total : curr_max
        if (height[left]<= height[right])
           left += 1
        else
            right -= 1
        end
    end
    curr_max
end

# Climbing Stairs

def climb_stairs(n)
    dp_array = Array.new(n)
    dp_array[0] = 1
    dp_array[1] = 1
    for i in (2..n)
       dp_array[i] = dp_array[i-1] + dp_array[i-2] 
    end
    dp_array[n]
end

#Tracks only two variables
def climb_stairs(n)
     return 1 if (n < 2)
    two_back = 1
    one_back = 1
    (n-1).times do |i|
        tmp = two_back
        two_back = one_back
        one_back = tmp + two_back
    end
    one_back
end

# coin change
def coin_change(coins, amount)
    sorted_coins = coins.sort
    dp = Array.new(amount+1, amount+1)
    dp[0] = 0
    for i in (1..amount)
       for j in (0...sorted_coins.length)
          coin_value = sorted_coins[j];
          break if (coin_value > amount)
           dp[i] = [dp[i], 1 + dp[i - coin_value]].min
       end
    end
    dp[amount] > amount ? -1 : dp[amount]
end

#Length of increasing subsequence
def length_of_lis(nums)
    result = [-1]
    current_largest = -1
    for i in (0...nums.length)
        inserted = false
        for j in (1...result.length)
            if (nums[i] <=  result[j])
                 result[j] = nums[i]
                 inserted = true
                 break
             end
          end
        result << nums[i] if !inserted
        end
    result.length - 1
end

require 'set'
def word_break(s, word_dict)
    wordDict = Set.new(word_dict)
    return true if wordDict.include?(s)
    is_valid = Array.new(s.length + 1, false)
    is_valid[0] = true;
    for i in (1..s.length)
       next if (!is_valid[i - 1]) 
        for j in (i..s.length)
            if (wordDict.include?(s[(i-1)...j]))
               is_valid[j] = true 
            end
        end
    end
    print is_valid
    is_valid[s.length]
end

def combination_sum4(nums, target)
    dp = Array.new(target + 1, 0)
    dp[0] = 1

    for i in (1..target)
       sum = 0
        nums.each do |num|
           curr = i - num
            next if curr < 0
            sum += dp[curr]
        end
        dp[i] = sum
    end
    dp[target]
    
end

def rob(nums)
    dp = Array.new(nums.length)
    dp[0] = nums[0]
    for i in (1...nums.length)
        if (i == 1)
            dp[i] = [nums[i], dp[i-1]].max
        else
            dp[i] = [nums[i] + dp[i-2], dp[i-1]].max 
        end
    end
    dp[nums.length - 1]
end